design-system/src/theme-provider.tsx

```diff
@@ -0,0 +1,68 @@
+import {
+  createContext,
+  useLayoutEffect,
+  useRef,
+  useState,
+  useMemo,
+  useContext,
+  useCallback,
+  type ReactNode,
+  type LegacyRef,
+} from 'react';
+import kebabCase from 'lodash/kebabCase';
+import { themes } from './custom-properties';
+
+type ThemeName = keyof typeof themes;
+
+const allThemesNames = Object.keys(themes);
+
+const ThemeContext = createContext({});
+
+const toVars = (obj: Record<string, string>) =>
+  Object.fromEntries(
+    Object.entries(obj).map(([key, value]) => [`--${kebabCase(key)}`, value])
+  );
+
+type ThemeProviderProps = {
+  children: ReactNode;
+};
+
+const ThemeProvider = (props: ThemeProviderProps) => {
+  const root = useRef<HTMLDivElement>();
+  const [theme, setTheme] = useState<ThemeName>('default');
+
+  const changeTheme = useCallback((newTheme: string) => {
+    if (allThemesNames.some((themeName) => themeName === newTheme)) {
+      setTheme(newTheme as ThemeName);
+    } else {
+      setTheme('default');
+    }
+  }, []);
+
+  useLayoutEffect(() => {
+    const vars = toVars(themes[theme]);
+    Object.entries(vars).forEach(([key, value]) => {
+      root.current?.style.setProperty(key, value);
```

Comment:
With this implementation, we don't remove the old theme vars from the document.
This is not a problem if we want themes to cascade. I mean, if we want to have the `default` one always loaded and then the rest to use it as a blueprint where they only overwrite specific parts.

There is an edge case though that I'm not sure it's worth taking into account:
* Let's say we have the these CSS vars in the `default` theme:
```
primaryColor: '#123123';
secondaryColor: '#234234';
```
* Then we have another theme (`theme-A`) where want to overwrite the primary color:
```
primaryColor: '#567567';
```
* And another theme (`theme-B`) where we overwrite the secondaryColor:
```
secondaryColor: '#789789';
```

If the users loads `default` theme, then `theme-A` and then `theme-B`, the resulting state would be:
```
primaryColor: '#567567';
secondaryColor: '#789789';
```
when maybe, as we have the `theme-B` selected and we want all themes to overwrite `default`, the state should be:
```
primaryColor: '#123123';
secondaryColor: '#789789';
```

It's just a matter of how we want the themes transition to work.

Reply:
I thought about this again and I think I was wrong.
Since we are inlining the styles, when we re-render the element with the new vars, the older ones would be gone.